pkg/tbtc/signing.go

```diff
@@ -233,3 +324,172 @@ func (srl *signingRetryLoop) qualifiedOperatorsSet() (
 
 	return chain.Addresses(qualifiedOperators).Set(), nil
 }
+
+// excludedMembersIndexes returns a list of excluded members' indexes for
+// the given qualified operators set.
+func (srl *signingRetryLoop) excludedMembersIndexes(
+	qualifiedOperatorsSet map[chain.Address]bool,
+	readyMembersIndexes []group.MemberIndex,
+) []group.MemberIndex {
+	includedMembersIndexes := make([]group.MemberIndex, 0)
+	excludedMembersIndexes := make([]group.MemberIndex, 0)
+	for i, operator := range srl.signingGroupOperators {
+		memberIndex := group.MemberIndex(i + 1)
+
+		if qualifiedOperatorsSet[operator] &&
+			slices.Contains(readyMembersIndexes, memberIndex) {
+			includedMembersIndexes = append(
+				includedMembersIndexes,
+				memberIndex,
+			)
+		} else {
+			excludedMembersIndexes = append(
+				excludedMembersIndexes,
+				memberIndex,
+			)
+		}
+	}
+
+	// Make sure we always use just the smallest required count of
+	// signing members for performance reasons
+	if len(includedMembersIndexes) > srl.chainConfig.HonestThreshold {
+		// #nosec G404 (insecure random number source (rand))
+		// Shuffling does not require secure randomness.
+		rng := rand.New(rand.NewSource(
+			srl.attemptSeed + int64(srl.attemptCounter),
+		))
+		// Sort in ascending order just in case.
+		sort.Slice(includedMembersIndexes, func(i, j int) bool {
+			return includedMembersIndexes[i] < includedMembersIndexes[j]
+		})
+		// Shuffle the included members slice to randomize the
+		// selection of additionally excluded members.
+		rng.Shuffle(len(includedMembersIndexes), func(i, j int) {
+			includedMembersIndexes[i], includedMembersIndexes[j] =
+				includedMembersIndexes[j], includedMembersIndexes[i]
+		})
+		// Get the surplus of included members and add them to
+		// the excluded members list.
+		excludedMembersIndexes = append(
+			excludedMembersIndexes,
+			includedMembersIndexes[srl.chainConfig.HonestThreshold:]...,
+		)
+		// Sort the resulting excluded members list in ascending order.
+		sort.Slice(excludedMembersIndexes, func(i, j int) bool {
+			return excludedMembersIndexes[i] < excludedMembersIndexes[j]
+		})
+	}
+
+	return excludedMembersIndexes
+}
+
+// broadcastSigningAnnouncer is an implementation of the signingAnnouncer that
+// performs the readiness announcement over the provided broadcast channel.
+type broadcastSigningAnnouncer struct {
```

Comment:
Worth considering extracting it to a separate file. Not necessarily in this PR but once we add the logic for operators joining the retries after the client restart.

Reply:
We can think about it but honestly, I think it's a good place for this type. The `signing.go` file contains a set of signing-related components that are closely related to each other. This is a quite Go-like way to organize things. We probably don't want a Java-like style where each significant type has its own file.